core/server/proxy/src/test/java/alluxio/proxy/s3/RateLimitInputStreamTest.java

```diff
@@ -0,0 +1,107 @@
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.FutureTask;
+import java.util.stream.Collectors;
+
+public class RateLimitInputStreamTest {
+
+  private byte[] mData;
+
+  @Before
+  public void init() throws IOException {
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream(MB);
+    int count = 0;
+    while (count < MB) {
+      byte[] bytes = UUID.randomUUID().toString().getBytes();
+      byteArrayOutputStream.write(bytes);
+      count += bytes.length;
+    }
+    mData = Arrays.copyOf(byteArrayOutputStream.toByteArray(), MB);
+  }
+
+  @Test
+  public void testSingleThreadRead() throws IOException {
+    long rate = 100 * KB;
+    long globalRate = 200 * KB;
+    ByteArrayInputStream inputStream = new ByteArrayInputStream(mData);
+    RateLimitInputStream rateLimitInputStream = new RateLimitInputStream(inputStream,
+        RateLimiter.create(rate), RateLimiter.create(globalRate));
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream(MB);
+    long start = System.currentTimeMillis();
+    IOUtils.copy(rateLimitInputStream, byteArrayOutputStream, KB);
+    long end = System.currentTimeMillis();
+    long duration = end - start;
+    long expectedDuration = MB / rate * 1000;
+    Assert.assertTrue(duration >= expectedDuration && duration <= expectedDuration + 1000);
+    Assert.assertArrayEquals(mData, byteArrayOutputStream.toByteArray());
+  }
+
+  @Test
+  public void testMultiThreadRead() {
+    int threadNum = 3;
+    long rate = 100 * KB;
+    long globalRate = 200 * KB;
+    long totalSize = (long) threadNum * mData.length;
+    RateLimiter globalRateLimiter = RateLimiter.create(globalRate);
+    ExecutorService threadPool = Executors.newFixedThreadPool(threadNum);
+    List<FutureTask<byte[]>> tasks = new ArrayList<>();
+    for (int i = 1; i <= threadNum; i++) {
+      tasks.add(new FutureTask<>(() -> {
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(mData);
+        RateLimitInputStream rateLimitInputStream = new RateLimitInputStream(inputStream,
+            RateLimiter.create(rate), globalRateLimiter);
+        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream(MB);
+        IOUtils.copy(rateLimitInputStream, byteArrayOutputStream, KB);
+        return byteArrayOutputStream.toByteArray();
+      }));
+    }
+    long start = System.currentTimeMillis();
+    tasks.forEach(threadPool::submit);
+    List<byte[]> results;
+    try {
+      results = tasks.stream().map(task -> {
+        try {
+          return task.get();
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      }).collect(Collectors.toList());
+    } finally {
+      threadPool.shutdownNow();
+    }
+    long end = System.currentTimeMillis();
+    long duration = end - start;
+    long expectedDuration = totalSize / globalRate * 1000;
```

Comment:
why expected is for the globalrate which is larger? the way RateLimitInputStream limits is to get both permits from both ratelimiter right, which means its always following the one with lower rate limit isnt it?

Reply:
The code has nothing to do with the order of ratelimiters, it will only limit the speed according to the minimum ratelimiter, i will change the test method.